### ISSUE TITLE
Bad email validator

### ISSUE THREAD

**Issue by iUnstable0** [Keats/validator#349]
This email validator incorrectly accepts `test@ok` as a valid input. While it follows the [HTML specification](https://html.spec.whatwg.org/multipage/forms.html#valid-e-mail-address) for email address validation, the regex provided on that website also matches emails without a top-level domain (TLD), such as `test@localhost`. This is problematic for several reasons:

1. **Public-Facing Applications**: For public-facing applications, accepting emails without a TLD can lead to invalid user inputs, causing potential issues with email delivery and user verification processes.

2. **Rare Usage in Development**: Even in development environments, it is rare to see local email addresses without a TLD. Most development setups use valid TLDs to simulate real-world scenarios.

To address these concerns, I suggest enforcing a stricter validation that requires a TLD. This can be achieved by modifying the regex to ensure the presence of a TLD. Here is an example of a stricter regex:

```regex
/^[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9-]+\.[a-zA-Z]{2,}$/
```

**Comment by Keats** [comment]
Honestly since this is allowed by both the mail spec (`test@gmailcom` is a valid email in theory) and the HTML5 spec I would prefer to leave it. However maybe we could have a strict mode that disable some esoteric addresses?

**Comment by iUnstable0** [comment]
itd be nice to have it built in! since right now i'm using custom function and the crate mailchecker

**Comment by szabgab** [comment]
I just encountered this problem as one of my users mistakenly used `blabla@gmail` forgetting the `.com`. I'd also like to see a way to make this stricter.

**Comment by GrantGryczan** [comment]
I agree, something like `#[validate(email_strict)]` would be nice. But with the way things typically go in Rust, I'd expect strict to be the default in order to avoid footguns, as I don't know if anyone really benefits from allowing TLDs. Either way, differing from the client-side validation for email inputs in HTML5 does arguably introduce a small problem: if the server is stricter than the client, then the user might see raw errors from the server rather than a more UX-friendly error (such as client-side validation). But I don't know that there's any way to prevent that footgun here, or if it's this crate's responsibility to anyway.

**Comment by Keats** [comment]
`strict` might not be a good name. More something like `allow_esoteric_mails`?

**Comment by szabgab** [comment]
What other cases are there besides "domain name without any dot in it"

As for name "require_one_dot_after_at"  or a more ambitious one: "valid_hostname" or "valid_tld" that would check against the list of all the TLDs.

**Comment by GrantGryczan** [comment]
> More something like `allow_esoteric_mails`?

The main purpose of it should be communicated by the name. I believe the purpose here is to act equivalently to how HTML5 validates emails. So I suspect something with `html5` in the name would be best, if you are taking the strict-by-default route.

**Comment by Keats** [comment]
The default doesn't need a name, it's the other one that is "simpler" that requires one

**Comment by GrantGryczan** [comment]
Oh, I thought you were suggesting a name for opting into the current default functionality, because this issue's suggested functionality is to _disallow_ certain emails from the default, not allow certain emails (unlike the impression I got from `allow_esoteric_emails`).

In that case, perhaps something like `#[validate(email(with_tld))]`? Again I think it's important for the name to communicate the purpose (so, for example, I don't have to check the docs periodically due to forgetting why it's needed).

**Comment by gianalarcon** [comment]
Hi team, any updates on this issue? We are also expecting to use a `validate(email_strict)` or something similar. Thanks!

**Comment by Keats** [comment]
I would take a PR for it

**Comment by GrantGryczan** [comment]
Haha, I just found out that some people actually do have real, working emails with just a TLD. See [this Reddit thread](https://www.reddit.com/r/programming/comments/8v60r/comment/c0aj5of/). In light of this, I'm not sure making this a built-in validator check is the right solution.

(Though I personally ended up going with the more ["Parse, don't validate"](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) solution of validating using newtypes with Serde, so I don't have any personal investment in this anymore.)

**Comment by cocoliliace** [comment]
> For public-facing applications, accepting emails without a TLD can lead to invalid user inputs, causing potential issues with email delivery and user verification processes.

I think this is a problem no parser can solve. Enforcing a stricter format would only hurt the usability.

Even addresses following the "strict" format can be non-existent or impersonations. If what you really want to verify is whether emails can be delivered to the address without issues, then the only way to verify it is to actually send a verification email to it. If the address is poorly formatted, the email would never get sent. Doing a premature syntax check has zero effect on the outcome if the address is invalid, but potentially negative effect if the address is valid but mistaken as invalid.
